fastcrypto-vdf/src/vdf/pietrzak/mod.rs

```diff
@@ -0,0 +1,150 @@
+    }
+}
+
+impl<G: ParameterizedGroupElement + Serialize> VDF for PietrzaksVDF<G> {
+    type InputType = G;
+    type OutputType = G;
+    type ProofType = Vec<G>;
+
+    fn evaluate(&self, input: &G) -> FastCryptoResult<(G, Vec<G>)> {
+        if !input.is_in_group(&self.group_parameter) || self.iterations == 0 {
+            return Err(InvalidInput);
+        }
+
+        // Compute output = 2^iterations * input
+        let output = input.repeated_doubling(self.iterations);
+
+        let mut x = input.clone();
+        let mut y = output.clone();
+        let mut t = self.iterations;
+
+        let mut proof = Vec::new();
+
+        // Compute the full proof. This loop may stop at any time which will give a shorter proof
+        // that is computationally harder to verify.
+        while t != 1 {
+            if check_parity_and_iterate(&mut t) {
+                y = y.double();
+            }
+
+            // TODO: Precompute some of the mu's to speed up the proof generation.
+            let mu = x.repeated_doubling(t);
+
+            let r = compute_challenge(&x, &y, self.iterations, &mu);
+            x = multiply(&x, &r, &self.group_parameter) + &mu;
+            y = multiply(&mu, &r, &self.group_parameter) + &y;
+
+            proof.push(mu);
+        }
+
+        Ok((output, proof))
+    }
+
+    fn verify(&self, input: &G, output: &G, proof: &Vec<G>) -> FastCryptoResult<()> {
+        if !input.is_in_group(&self.group_parameter)
+            || !output.is_in_group(&self.group_parameter)
+            || !proof.iter().all(|mu| mu.is_in_group(&self.group_parameter))
+            || self.iterations == 0
+        {
+            return Err(InvalidInput);
+        }
+
+        let mut x = input.clone();
+        let mut y = output.clone();
+        let mut t = self.iterations;
+
+        for mu in proof {
+            if check_parity_and_iterate(&mut t) {
+                y = y.double();
+            }
+
+            let r = compute_challenge(&x, &y, self.iterations, mu);
+            x = multiply(&x, &r, &self.group_parameter) + mu;
+            y = multiply(mu, &r, &self.group_parameter) + y;
+        }
+
+        // In case the proof is shorter than the full proof, we need to compute the remaining powers.
+        x = x.repeated_doubling(t);
+        if x != y {
+            return Err(InvalidProof);
+        }
+        Ok(())
+    }
+}
+
+/// Compute the Fiat-Shamir challenge used in Pietrzak's VDF construction.
+fn compute_challenge<G: ParameterizedGroupElement + Serialize>(
+    input: &G,
+    output: &G,
+    iterations: u64,
+    mu: &G,
+) -> BigUint {
+    let seed = bcs::to_bytes(&(input, output, iterations, mu))
+        .expect("Failed to serialize Fiat-Shamir input.");
+    let hash = Keccak256::digest(seed);
+    BigUint::from_bytes_be(&hash.digest[..DEFAULT_CHALLENGE_SIZE_IN_BYTES])
+}
+
+/// Replace t with (t+1) >> 1 and return true iff the input was odd.
+#[inline]
+fn check_parity_and_iterate(t: &mut u64) -> bool {
```

Comment:
assert that t < MAX_U64

Reply:
I feel that the current version is still risky in case the adversary controls t. can we gracefully handle MAX_U64?

Reply:
We can use the fact that (t + 1) >> 1 = t >> 1 + t%2 to avoid potentially overflowing addition.